source/itertools/index.rst

```diff
@@ -6,183 +6,133 @@
-avec une simple fonction :py:func:`count() <itertools.count()>` ou par exemple
-créer un filtre sur nos tableau avec :py:func:`compress<itertools.compress>`.
-
-
-Itérateur infini
-----------------
-
-fonction :py:func:`count(start=0, step=1) <itertools.count>`
-
-Fonction :py:func:`itertools.count` est un itérateur infini. Il prend comme
-paramètres un début et le *step* (par défault c'est 1).
-
-.. literalinclude:: use_itertools.py
-   :start-after: # COUNT_BEGIN
-   :end-before: # COUNT_END
-
+.. code-block:: pycon
 
-Itérateur avec fin
--------------------
+>>> liste = ["I", "do" "love", "programming", "in", "python"]
+>>> for i in liste:
+    print(i)
+I
+do
+love
+programming
+in
+python
 
-fonction :py:func:`chain() <itertools.chain>`
-*********************************************
+Derrière cette boucle **for** se cache enfet un itérateur. C'est un objet qui va être chargé de parcourir un objet conteneur, dans  notre cas notre liste. Quand Python va tomber sur la ligne **for i in liste:**, il va appeler l'itérateur de notre liste pour pouvoir la parcourir. L'itérateur se crée dans la méthode **__iter__** de notre objet liste dans notre cas. Et lui va nous retourner notre itérateur pour pouvoir parcourir notre petit liste.
 
-Chain permet de concaténer des listes ensembles en lui passant en paramètre
-les listes que vous voulez concatener.
+A chaque itération python va appeler la méthode **__next__** de notre liste pour itérer à l'élément suivant ou s'arrêter avec l'execption **StopIteration** si le parcours à fini de parcour la liste.
 
-Par exemple :
+Voici un autre exemple mais avec une chaine string.
 
 .. code-block:: pycon
 
-    >>> a = [1, 2, 3, 4, 5, 6, 7]
-    >>> b = ["Johnny", "David", "Mike", "Bali", "Noami"]
-    >>> c = ["c", "h", "b", "s", "z"]
-    >>> for i in chain(a, b, c):
-        print(i)
-    1
-    2
-    3
-    4
-    5
-    6
-    7
-    Johnny
-    David
-    Mike
-    Bali
-    Noami
-    c
-    h
-    b
-    s
-    z
-
-.. todo::
-
-    ``list`` n'est pas une fonction mais un constructeur.
-
-Une façon plus élégante d'afficher une liste est d'utiliser 
-la fonction list qui va nous retourner une liste(captain obvious) qui vous
-suffira de :py:func:`print`.
-
-.. code-block:: pycon
+>>> chaine = "Python"
+>>> iterateur = iter(chaine) # va nous retourner un itérateur sur notre chaine
+>>> print(next(iterateur)) # va nous afficher la première lettre de notre string
+P
+>>> for i in chaine: # va parcourir tous les caractère un à un
+    print(i)
+P
+y
+t
+h
+o
+n
 
-    >>> print(list(chain(a, b, c)))
-    [1, 2, 3, 4, 5, 6, 7, 'Johnny', 'David', 'Mike', 'Bali', 'Noami', 'c', 'h', 'b', 's', 'z']
-    >>> print(list(chain(a[:1], b[:1], c[:1])))
-    [1, 'Johnny', 'c']
-    >>> print(list(chain(a[0:2], b[0:2], c[0:2]))
-    [1, 2, 'Johnny', 'David', 'c', 'h']
+Quand au générateurs, ce sont des outils puissants pour créer et manipuler des itérateurs ce qui permet de créer des choses complexe avec très peu de code.
```

Comment:
Voici un générateur (à cause du `yield`). De manière simpliste, c'est une fonction qui peut être stoppée et redémarrée.

```python
def count():
    i = 0
    while True:
        yield i
        i += 1
```

Voici un itérateur.

```python
i = iter(count())
try:
    while True:
        val = next(i)
        # ...
raise StopIteration:
    pass
```

Un itérateur, selon de _design pattern_ est simplement un objet permettant de pacourir une liste, finie ou infinie. L'expression `for` n'est rien d'autre que la bouche `ẁhile` ci-dessus.

`itertools` est un module permettant de faire mumuse avec cet objet très simple qu'est l'itérateur.

Générateur et itérateur sont intimement liés.